modules/ml/include/opencv2/ml.hpp

```diff
@@ -1499,6 +1502,34 @@ class CV_EXPORTS_W ANN_MLP : public StatModel
     /** @copybrief getRpropDWMax @see getRpropDWMax */
     CV_WRAP virtual void setRpropDWMax(double val) = 0;
 
+    /** ANNEAL: Update initial temperature.
+    It must be \>=0. Default value is 10.*/
+    /** @see setAnnealInitialT */
+    CV_WRAP virtual double getAnnealInitialT() const = 0;
```

Comment:
There is no way to alter virtual function table without breaking of library ABI interface.

Try using this:
```
// .hpp file
class CV_EXPORTS_W ANN_MLP : public StatModel
{
    ...
    CV_WRAP double getAnnealInitialT() const = 0; // <-- no virtual;
    ... other new methods ...
}

class CV_EXPORTS_W ANN_MLP_ANNEAL : public ANN_MLP
{
    ... new virtual methods ...
}

//ann_mlp.cpp

double ANN_MLP::getAnnealInitialT() const
{
    const ANN_MLP_ANNEAL* this_ = dynamic_cast<const ANN_MLP_ANNEAL*>(this);
    if (!this_)
        CV_Error(Error::StsNotImplemented, "the class is not ANN_MLP_ANNEAL");
    return this_->getAnnealInitialT();
}

//replace class ANN_MLPImpl : public ANN_MLP
class ANN_MLPImpl : public ANN_MLP_ANNEAL
```

Also there is option without "ANN_MLP_ANNEAL" class, but "dynamic_cast" is used with "ANN_MLPImpl" directly.

Reply:
I try and I must copy all methods in ANN_MLPImpl isn't it?

Reply:
All **new** methods.

Reply:
In ann_mlp.cpp I insert: 
class ANN_MLP_ANNEALImpl : public ANN_MLP_ANNEAL
{
public:
    ANN_MLP_ANNEALImpl()
    {
        clear();
        setActivationFunction(SIGMOID_SYM, 0, 0);
        setLayerSizes(Mat());
        setTrainMethod(ANN_MLP::ANNEAL, 0.1, FLT_EPSILON, 0, 0);
    }
    virtual ~ANN_MLP_ANNEALImpl() {}
        CV_IMPL_PROPERTY(double, AnnealInitialT, params.initialT)
        CV_IMPL_PROPERTY(double, AnnealFinalT, params.finalT)
        CV_IMPL_PROPERTY(double, AnnealCoolingRatio, params.coolingRatio)
        CV_IMPL_PROPERTY(int, AnnealItePerStep, params.itePerStep)
        
     AnnParams params;
};

Ptr<ANN_MLP_ANNEAL> ANN_MLP_ANNEAL::create()
{
    return makePtr<ANN_MLP_ANNEALImpl>(); <---- ERROR  2259
}


>G:\Lib\opencv\modules\core\include\opencv2/core/ptr.inl.hpp(301): error C2259: 'cv::ml::ANN_MLP_ANNEALImpl' : class' : cannot instantiate abstract class
see  G:\Lib\opencv\modules\core\include\opencv2/core/ptr.inl.hpp(301): note: 'int cv::ml::StatModel::getVarCount(void) const' :  is abstract

and all methods are enumerated


My anneal class is
class CV_EXPORTS_W ANN_MLP_ANNEAL : public ANN_MLP
{
public:
    /** ANNEAL: Update initial temperature.
    It must be \>=0. Default value is 10.*/
    /** @see setAnnealInitialT */
    CV_WRAP virtual double getAnnealInitialT() const;
    /** @copybrief getAnnealInitialT @see getAnnealInitialT */
    CV_WRAP virtual void setAnnealInitialT(double val) const;

    /** ANNEAL: Update final temperature.
    It must be \>=0 and less than initialT. Default value is 0.1.*/
    /** @see setAnnealFinalT */
    CV_WRAP  virtual double getAnnealFinalT() const;
    /** @copybrief getAnnealFinalT @see getAnnealFinalT */
    CV_WRAP  virtual void setAnnealFinalT(double val) ;

    /** ANNEAL: Update cooling ratio.
    It must be \>0 and less than 1. Default value is 0.95.*/
    /** @see setAnnealCoolingRatio */
    CV_WRAP  virtual double getAnnealCoolingRatio() const ;
    /** @copybrief getAnnealCoolingRatio @see getAnnealCoolingRatio */
    CV_WRAP  virtual void setAnnealCoolingRatio(double val) const;

    /** ANNEAL: Update iteration per step.
    It must be \>0 . Default value is 10.*/
    /** @see setAnnealItePerStep */
    CV_WRAP virtual int getAnnealItePerStep() const;
    /** @copybrief getAnnealItePerStep @see getAnnealItePerStep */
    CV_WRAP virtual  void setAnnealItePerStep(int val) const;


    /** @brief Creates empty model

    Use StatModel::train to train the model, Algorithm::load\<ANN_MLP\>(filename) to load the pre-trained model.
    Note that the train method has optional flags: ANN_MLP::TrainFlags.
    */
    CV_WRAP static Ptr<ANN_MLP_ANNEAL> create();

    /** @brief Loads and creates a serialized ANN from a file
    *
    * Use ANN::save to serialize and store an ANN to disk.
    * Load the ANN from this file again, by calling this function with the path to the file.
    *
    * @param filepath path to serialized ANN
    */
    CV_WRAP static Ptr<ANN_MLP_ANNEAL> load(const String& filepath);
....

Reply:
> In ann_mlp.cpp I insert

Need just replace base class `ANN_MLP` => `ANN_MLP_ANNEAL` of `ANN_MLPImpl` (not `ANN_MLP_ANNEALImpl`).